File: Task2/Task2_1.js
// Prepare promise based alternative
// Write use cases for the promise based solution

function promiseMap(array, callback) {
    const promises = array.map(item => callback(item));
    return Promise.all(promises);
}

async function demoFunc() {

    
    // case 1
    const numbers = [1, 2, 3, 4, 5];

    const promiseTriple = (num) => {
        return new Promise((resolve) => {
            setTimeout(() => resolve(num * 3), Math.random() * 1000);
        });
    };

    promiseMap(numbers, promiseTriple)
        .then(results => {
            console.log(results);
        });

    
    //case 2
    const numbers2 = [10, 20, 30];

    const promiseSquare = (num) => {
        return new Promise((resolve) => {
            setTimeout(() => resolve(num * num), Math.random() * 1000);
        });
    };
    
    promiseMap(numbers2, promiseSquare)
    .then(squares => {
        console.log(squares);
    });
}

demoFunc();

File: Task2/Task2_1.kt
// Prepared promise based alternative
// Wrote use case for the promise based solution

import kotlinx.coroutines.*

fun <T, R> promiseMap(scope: CoroutineScope, list: List<T>, callback: (T) -> Deferred<R>): Deferred<List<R>> {
    return scope.async {
        list.mapNotNull {
            try {
                callback(it).await()
            } catch (e: Exception) {
                println("ERROR! ${e.message}")
                null
            }
        }
    }
}

fun main() = runBlocking {
    val numbers = listOf(1, 2, 3, 4, 5)

    val promiseTriple: (Int) -> Deferred<Int> = { num ->
        CompletableDeferred<Int>().apply {
            launch {
                delay(1000)
                if (num == 3) {
                    completeExceptionally(Exception("Cannot process number $num"))
                } else {
                    complete(num * 3)
                }
            }
        }
    }
    
    val results = promiseMap(this, numbers, promiseTriple).await()
    println(results)
}
